# Exploración de datos: 
## Centros educativos autorizados por el Mineduc y carreras que pueden impartir.

### Librerías

In [1]:
# librerías
import pandas as pd
import numpy as np

### Cargar

In [2]:
df = pd.read_csv("establecimientos.csv", encoding="utf-8-sig")
df.head()

,CODIGO,DISTRITO,DEPARTAMENTO,MUNICIPIO,ESTABLECIMIENTO,DIRECCION,TELEFONO,SUPERVISOR,DIRECTOR,NIVEL,SECTOR,AREA,STATUS,MODALIDAD,JORNADA,PLAN,DEPARTAMENTAL
0,01-01-0007-46,01-504,GUATEMALA,GUATEMALA,COLEGIO HAN AL AMERICANO,9A. CALLE 28-74 ZONA 14,23661344-23661600,DAVID SOTOJ SANCHEZ,OLGA LETICIA PEREZ GARICA,DIVERSIFICADO,PRIVADO,URBANA,CERRADA DEFINITIVAMENTE,MONOLINGUE,MATUTINA,DIARIO(REGULAR),GUATEMALA NORTE
1,01-01-0011-46,01-620,GUATEMALA,GUATEMALA,COLEGIO LA EDUCACION,7A. AVENIDA 3-20 ZONA 19 COLONIA LA FLORIDA,24310333,RONALD MAURICIO MORALES VIVAR,NATALY GABRIELA DE LEON URÍZA,DIVERSIFICADO,PRIVADO,URBANA,CERRADA DEFINITIVAMENTE,MONOLINGUE,VESPERTINA,DIARIO(REGULAR),GUATEMALA NORTE
2,01-01-0013-46,01-503,GUATEMALA,GUATEMALA,COLEGIO LA ACADEMIA,13 CALLE 6-91 ZONA 10,23323279,NANCY CAROLINA NAVAS GONZÁLEZ,ANA LUCRECIA LOPEZ FLORES,DIVERSIFICADO,PRIVADO,URBANA,CERRADA DEFINITIVAMENTE,MONOLINGUE,MATUTINA,DIARIO(REGULAR),GUATEMALA NORTE
3,01-01-0015-46,01-403,GUATEMALA,GUATEMALA,"CENTRO EDUCATIVO ""RENUEVO""",6A. 2-34 ZONA 1,55856196,CARLOS HUMBERTO GONZÁLEZ DE LEÓN,NaN,DIVERSIFICADO,PRIVADO,RURAL,CERRADA DEFINITIVAMENTE,MONOLINGUE,INTERMEDIA,FIN DE SEMANA,GUATEMALA NORTE
4,01-01-0018-46,01-307,GUATEMALA,GUATEMALA,LICEO ROMANO AMERICANO,6A. CALLE 2-12 ZONA 1,NaN,NORMA ARACELY PALOMO FRANCO DE DIAZ,VICTOR MANUEL PORTILLO LEON,DIVERSIFICADO,PRIVADO,URBANA,CERRADA TEMPORALMENTE,MONOLINGUE,MATUTINA,DIARIO(REGULAR),GUATEMALA NORTE


In [3]:
pd.DataFrame(df, columns=['DEPARTAMENTO', 'STATUS', 'NIVEL']).head() #es case sensitive

,DEPARTAMENTO,STATUS,NIVEL
0,GUATEMALA,CERRADA DEFINITIVAMENTE,DIVERSIFICADO
1,GUATEMALA,CERRADA DEFINITIVAMENTE,DIVERSIFICADO
2,GUATEMALA,CERRADA DEFINITIVAMENTE,DIVERSIFICADO
3,GUATEMALA,CERRADA DEFINITIVAMENTE,DIVERSIFICADO
4,GUATEMALA,CERRADA TEMPORALMENTE,DIVERSIFICADO


## 3. Análisis del estado inicial de los datos
#### a. Número de registros y variables. 



In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9728 entries, 0 to 9727
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   CODIGO           9706 non-null   str  
 1   DISTRITO         9405 non-null   str  
 2   DEPARTAMENTO     9706 non-null   str  
 3   MUNICIPIO        9706 non-null   str  
 4   ESTABLECIMIENTO  9703 non-null   str  
 5   DIRECCION        9637 non-null   str  
 6   TELEFONO         9024 non-null   str  
 7   SUPERVISOR       9402 non-null   str  
 8   DIRECTOR         8510 non-null   str  
 9   NIVEL            9706 non-null   str  
 10  SECTOR           9706 non-null   str  
 11  AREA             9706 non-null   str  
 12  STATUS           9706 non-null   str  
 13  MODALIDAD        9706 non-null   str  
 14  JORNADA          9706 non-null   str  
 15  PLAN             9706 non-null   str  
 16  DEPARTAMENTAL    9706 non-null   str  
dtypes: str(17)
memory usage: 1.3 MB


El dataset contiene 17 variables y 9,728 registros que proveen de información acerca de todos los centros educativos según el MINEDUC.

#### b. Tipo de dato de cada variable.



In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9728 entries, 0 to 9727
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   CODIGO           9706 non-null   str  
 1   DISTRITO         9405 non-null   str  
 2   DEPARTAMENTO     9706 non-null   str  
 3   MUNICIPIO        9706 non-null   str  
 4   ESTABLECIMIENTO  9703 non-null   str  
 5   DIRECCION        9637 non-null   str  
 6   TELEFONO         9024 non-null   str  
 7   SUPERVISOR       9402 non-null   str  
 8   DIRECTOR         8510 non-null   str  
 9   NIVEL            9706 non-null   str  
 10  SECTOR           9706 non-null   str  
 11  AREA             9706 non-null   str  
 12  STATUS           9706 non-null   str  
 13  MODALIDAD        9706 non-null   str  
 14  JORNADA          9706 non-null   str  
 15  PLAN             9706 non-null   str  
 16  DEPARTAMENTAL    9706 non-null   str  
dtypes: str(17)
memory usage: 1.3 MB


Por la manera en la que se importaron los datos fue imposible conservar el formato o tipo original de cada variable, por lo que todas son reconocidas por Pandas como string. 

Sin embargo, es evidente que la variable, TELEFONO, es de tipo entero con 8 dígitos y CODIGO es un string que podría ser dividido en tres variables obtenidas al separar los números por guión.

#### c. Cantidad y porcentaje de valores faltantes por variable.



In [6]:
valores_faltantes = df.isnull().sum()

# 2 promedio
porcentaje_faltantes = df.isnull().mean() * 100

# tabla
tabla_faltantes = pd.DataFrame({
    'Cantidad Faltantes': valores_faltantes,
    'Porcentaje (%)': porcentaje_faltantes
})
tabla_faltantes

#Ordenado por porcentaje de faltantes
tabla_faltantes = tabla_faltantes.sort_values(by="Porcentaje (%)", ascending=False)
tabla_faltantes


,Cantidad Faltantes,Porcentaje (%)
DIRECTOR,1218,12.520559
TELEFONO,704,7.236842
SUPERVISOR,326,3.351151
DISTRITO,323,3.320312
DIRECCION,91,0.935444
ESTABLECIMIENTO,25,0.256990
CODIGO,22,0.226151
MUNICIPIO,22,0.226151
DEPARTAMENTO,22,0.226151
NIVEL,22,0.226151


Alrededor del 13% de los registros de DIRECTOR contienen datos faltantes, lo cual sugiere que existen falencias en el registro de datos debido a que cualquier instituto en Guatemala ha de tener un director. Según la Ley de Educación Nacional de Guatemala (Decreto 12-91), todo centro educativo, ya sea público, privado o por cooperativa, debe contar con un director. Esta figura es la máxima autoridad del establecimiento y la responsable legal, administrativa y pedagógica ante el Ministerio de Educación (MINEDUC).

Por otra parte, no se cuenta con un 7% de los teléfonos de todos los institutos. 

Un 3% de los centros educativos carecen de supervisor, lo cual es posible si se refiere a un supervisor interno. Ahora, si esta variable está destinada para describir al Supervisor Educativo asignado por el Ministerio de Educación, entonces se trata de datos faltantes.

Dirección, distrito y establecimiento presentan menos del 4% de información faltante, del que se sugiere que una parte de ella se debe a que existen 22 registros en donde no se cuenta con información alguna, que posiblemente se debe a espacios entre filas.

Las demás variables tienen una ínfima cantidad de datos faltantes y parece que se trata acerca de estas filas fantasma que no contienen información en ninguna variable.


#### d. Cantidad de valores únicos.



In [7]:
df.nunique()
#9706*100/9728

CODIGO             9706
DISTRITO           1593
DEPARTAMENTO         22
MUNICIPIO           330
ESTABLECIMIENTO    5036
DIRECCION          6023
TELEFONO           5547
SUPERVISOR         1241
DIRECTOR           4852
NIVEL                 1
SECTOR                4
AREA                  3
STATUS                5
MODALIDAD             2
JORNADA               6
PLAN                 13
DEPARTAMENTAL        26
dtype: int64

La variable, CODIGO, está compuesta por un 99.77% de registros únicos; seguido por DIRECCION y TELEFONO como otras variables con mayor cantidad de valores únicos. La variable d DISTRITO contiene 1593. Se esperaba obtener 22 departamentos y municipios solamente hay 330 según el dataset (donde el total oficial es de 340, por lo que 10 municipios ya sea que no cuenten con institutos de diversificado o no se haya registrado correctamente ese registro). 

En total existen 13 planes educativos, 6 tipos de jornadas, 2 modalidades, hay 5 status distintos, 3 áreas, 4 sectores y solamente 1 nivel educativo (diversificado) por la restricción implementada al importar datos.

#### e. Cantidad de registros duplicados exactos.



In [8]:
resumen_duplicados = pd.DataFrame({
    'Métrica': [
        'Duplicados exactos (todas las ocurrencias, keep=False)',
        'Duplicados exactos (sin la primera ocurrencia)',
        'Filas completamente vacías (todas las columnas nulas)',
        'Duplicados exactos entre registros CON datos',
        'Códigos (CODIGO) repetidos entre registros no nulos',
    ],
    'Cantidad': [
        df.duplicated(keep=False).sum(),
        df.duplicated().sum(),
        df.isnull().all(axis=1).sum(),
        df[~df.isnull().all(axis=1)].duplicated().sum(),
        df[df['CODIGO'].notna()].duplicated(subset=['CODIGO']).sum(),
    ]
})
resumen_duplicados

,Métrica,Cantidad
0,"Duplicados exactos (todas las ocurrencias, kee...",22
1,Duplicados exactos (sin la primera ocurrencia),21
2,Filas completamente vacías (todas las columnas...,22
3,Duplicados exactos entre registros CON datos,0
4,Códigos (CODIGO) repetidos entre registros no ...,0


duplicated() marca 21 registros como duplicados exactos (22 si se cuentan todas las ocurrencias con keep=False). Sin embargo, al revisarlos se determinó que la totalidad corresponde a filas completamente vacías (sin ningún valor en sus 17 columnas): son las mismas 22 filas responsables de los valores faltantes reportados en el inciso c, lo cual explica por qué la mayoría de las variables comparten un 0.23% de nulos.

Al excluir esas filas vacías, no queda ningún duplicado exacto entre establecimientos con información real, y la variable CODIGO, que actúa como identificado, no presenta ningún valor repetido entre los registros no nulos. Este hallazgo también implica que las filas vacías deberán eliminarse en el plan de limpieza, ya que están inflando artificialmente tanto el conteo de faltantes como el de duplicados.

#### f. Variables con valores fuera de dominio o inconsistentes.



In [9]:
df[['AREA', 'STATUS']].apply(lambda s: s.value_counts(dropna=False))

,AREA,STATUS
ABIERTA,NaN,6038.0
CERRADA DEFINITIVAMENTE,NaN,1286.0
CERRADA TEMPORALMENTE,NaN,2291.0
RURAL,2391.0,NaN
SIN ESPECIFICAR,2.0,NaN
TEMPORAL NOMBRAMIENTO,NaN,47.0
TEMPORAL TITULOS,NaN,44.0
URBANA,7313.0,NaN
NaN,22.0,22.0


In [10]:
telefonos = df['TELEFONO'].dropna().astype(str)
codigos = df['CODIGO'].dropna()

pd.DataFrame({
    'Variable': ['TELEFONO', 'CODIGO'],
    'Regla esperada': ['8 dígitos numéricos', 'NN-NN-NNNN-NN'],
    'No cumplen': [(~telefonos.str.fullmatch(r'\d{8}')).sum(), (~codigos.str.fullmatch(r'\d{2}-\d{2}-\d{4}-\d{2}')).sum()],
    'Total evaluado': [len(telefonos), len(codigos)]
})

,Variable,Regla esperada,No cumplen,Total evaluado
0,TELEFONO,8 dígitos numéricos,163,9024
1,CODIGO,NN-NN-NNNN-NN,0,9706


AREA incluye la categoría "SIN ESPECIFICAR" (2 registros), fuera del dominio binario esperado {URBANA, RURAL}. STATUS presenta dos categorías poco frecuentes (TEMPORAL NOMBRAMIENTO y TEMPORAL TITULOS, con 47 y 44 registros) que son válidas pero merecen revisión por su baja representación.

La variable con más problemas de dominio es TELEFONO: los registros contienen caracteres no numéricos (guiones, comas o texto), y entre los valores puramente numéricos hay longitudes de 1, 2, 4, 5, 6, 7, 9, 10 y hasta 16 dígitos en lugar de los 8 dígitos esperados para un teléfono guatemalteco.

En contraste, DEPARTAMENTO (validado en el inciso d) y CODIGO sí cumplen su dominio esperado: los 22 valores de DEPARTAMENTO coinciden exactamente con los 22 departamentos oficiales de Guatemala, y el 100% de los códigos no nulos siguen el patrón NN-NN-NNNN-NN.

#### g. Variables con formatos inconsistentes.



In [11]:
est = df['ESTABLECIMIENTO'].dropna().astype(str)
dire = df['DIRECCION'].dropna().astype(str)
distrito = df['DISTRITO'].dropna().astype(str)
formato_distrito = distrito.str.fullmatch(r'\d{2}-\d{3}') | distrito.str.fullmatch(r'\d{2}-\d{2}-\d{4}')

pd.DataFrame({
    'Variable': ['ESTABLECIMIENTO', 'ESTABLECIMIENTO', 'DIRECCION', 'TELEFONO', 'DISTRITO'],
    'Problema de formato': [
        'Usa comillas dobles', 'Usa comillas simples', 'Contiene minúsculas',
        'Contiene guion, coma o texto', 'No sigue NN-NNN ni NN-NN-NNNN',
    ],
    'Cantidad': [
        est.str.contains(chr(34)).sum(),
        est.str.contains(chr(39)).sum(),
        dire.str.contains('[a-z]', regex=True).sum(),
        (~telefonos.str.fullmatch(r'\d+')).sum(),
        (~formato_distrito).sum(),
    ]
})

,Variable,Problema de formato,Cantidad
0,ESTABLECIMIENTO,Usa comillas dobles,2028
1,ESTABLECIMIENTO,Usa comillas simples,508
2,DIRECCION,Contiene minúsculas,7
3,TELEFONO,"Contiene guion, coma o texto",123
4,DISTRITO,No sigue NN-NNN ni NN-NN-NNNN,16


Además de los problemas de dominio, TELEFONO mezcla distintos formatos dentro de la misma columna: números separados por guión, por coma o acompañados de texto, lo que indica que el campo a veces almacena más de un contacto en lugar de un solo teléfono.

ESTABLECIMIENTO no tiene un estilo uniforme para encerrar apodos o nombres comerciales: 2,028 registros usan comillas dobles y 508 usan comillas simples. DIRECCION es casi enteramente uniforme en mayúsculas (solo 7 de 9,637 registros contienen alguna letra minúscula), pero esos mismos registros muestran una abreviatura inconsistente del ordinal de calle/avenida (p. ej. "8a. AVE." en minúscula frente a la convención dominante "8A. AVE."), evidenciando que la normalización de texto tendrá que estandarizar mayúsculas y el uso de puntos en las abreviaturas.

DISTRITO también mezcla dos formatos estructurales distintos: 3,928 registros siguen el patrón NN-NNN (p. ej. "01-406") y 5,461 siguen NN-NN-NNNN (p. ej. "01-01-0146"), además de 16 valores truncados sin dígitos después del guion (p. ej. "01-"). Esto sugiere que los datos provienen de al menos dos convenciones de codificación distintas y deberá decidirse a cuál estandarizar antes de usar esta variable.

#### h. Identificación de problemas potenciales de calidad de datos.



In [12]:
import unicodedata
normalizar = lambda s: ''.join(c for c in unicodedata.normalize('NFD', str(s)) if unicodedata.category(c) != 'Mn').upper()

# DEPARTAMENTAL subdivide GUATEMALA (4 zonas) y QUICHE (2); validamos que
# siempre correspondan al DEPARTAMENTO correcto (consistencia entre variables)
sub = df[['DEPARTAMENTO', 'DEPARTAMENTAL']].dropna()
dep_norm, depal_norm = sub['DEPARTAMENTO'].map(normalizar), sub['DEPARTAMENTAL'].map(normalizar)
coincide = [d.startswith(dep) for d, dep in zip(depal_norm, dep_norm)]

pd.DataFrame({
    'Métrica': ['Registros comparados', 'Contradicciones'],
    'Cantidad': [len(sub), (~pd.Series(coincide)).sum()]
})

,Métrica,Cantidad
0,Registros comparados,9706
1,Contradicciones,0


Al cruzar DEPARTAMENTO con DEPARTAMENTAL no se encontró ninguna contradicción: las 26 categorías de DEPARTAMENTAL siempre corresponden al departamento correcto en DEPARTAMENTO. Es una validación cruzada exitosa entre dos variables relacionadas, y descarta que la granularidad extra de DEPARTAMENTAL (visible en el inciso d como "26 valores únicos" frente a los 22 departamentos) sea un error de inconsistencia; es, en realidad, una subdivisión administrativa válida (Guatemala en 4 zonas, Quiché en 2).

## 4. Plan de limpieza

Antes de modificar los datos, se define a continuación un plan de limpieza para cada variable, basado en los hallazgos del diagnóstico (inciso 3). Como paso previo y transversal a todas las variables, se eliminarán las **22 filas completamente vacías** (inciso 3e). Las cantidades de "problemas encontrados" de la tabla ya excluyen esas 22 filas cuando corresponde, para no contar el mismo problema dos veces.

| Variable | a. Problemas encontrados | b. Regla de corrección | c. Riesgos |
|---|---|---|---|
| **CODIGO** | 22 faltantes = filas vacías. Formato ya consistente. | Eliminar filas vacías; mantener como texto. | Perder un registro si la fila no está realmente vacía. |
| **DISTRITO** | 301 faltantes; 2 formatos mezclados (NN-NNN / NN-NN-NNNN); 16 truncados. | Detectar el patrón por regex y unificar a NN-NN-NNNN; descartar truncados. | No se sabe si ambos formatos son equivalentes; se puede perder trazabilidad. |
| **DEPARTAMENTO** | 22 faltantes; dominio limpio. | Solo eliminar/documentar vacíos. | Ninguno. |
| **MUNICIPIO** | 22 faltantes; 330 de 340 municipios oficiales presentes, sin validar contra catálogo ni DEPARTAMENTO. | Normalizar texto y cruzar contra catálogo oficial por departamento. | Nombres se repiten entre departamentos; riesgo de reasignar mal. |
| **ESTABLECIMIENTO** | 3 faltantes; comillas mezcladas (dobles/simples). | Unificar a comillas dobles. | Cambio cosmético; documentar que no altera el nombre real. |
| **DIRECCION** | 69 faltantes; 7 con minúsculas y abreviaturas de ordinal inconsistentes. | Mayúsculas + regex para normalizar ordinales ("8a." → "8A."). | La regex puede alterar texto libre no relacionado (nombres de colonias). |
| **TELEFONO** | 682 faltantes; 123 con texto/guiones/comas; longitudes de 2 a 26 caracteres. | Extraer el primer número de 8 dígitos; mover el resto a TELEFONO_2; NA si no calza. | Mayor riesgo del proyecto: se puede descartar el número más útil. |
| **SUPERVISOR** | 304 faltantes; posibles variantes de nombre sin verificar. | Solo señalar duplicados por similitud (RapidFuzz), sin fusionar. | Tocayos distintos podrían fusionarse por error. |
| **DIRECTOR** | 1,196 faltantes (la más alta); mismo riesgo que SUPERVISOR. | Igual que SUPERVISOR; documentar que el faltante puede ser vacante real. | Asumir que falta = error de captura sería incorrecto. |
| **NIVEL** | 22 faltantes; dominio ya limpio. | Solo eliminar/documentar vacíos. | Ninguno. |
| **SECTOR** | 22 faltantes; dominio limpio. | Solo documentar/eliminar vacíos. | Ninguno. |
| **AREA** | 22 faltantes; "SIN ESPECIFICAR" fuera de dominio (2 casos). | Recodificar "SIN ESPECIFICAR" como NA. | Podría significar "no aplica", no un faltante real. |
| **STATUS** | 22 faltantes; categorías raras pero válidas. | Sin cambios; documentar las categorías raras. | Podrían agruparse por error con las comunes. |
| **MODALIDAD** | 22 faltantes; dominio limpio. | Solo documentar/eliminar vacíos. | Ninguno. |
| **JORNADA** | 22 faltantes; "SIN JORNADA" sin verificar si es real. | Cruzar con PLAN para confirmar antes de recodificar. | Podría borrarse información válida de planes a distancia. |
| **PLAN** | 22 faltantes; categorías raras (pocos casos). | Sin cambios; documentar rareza, no fusionar. | Fusionar perdería granularidad relevante. |
| **DEPARTAMENTAL** | 22 faltantes; 100% consistente con DEPARTAMENTO. | Sin cambios; documentar como subdivisión administrativa. | Ninguno. |